src/App.tsx

```diff
@@ -1,14 +1,67 @@
 /* eslint-disable max-len */
-import React from 'react';
+import React, { useState, useEffect, useCallback } from 'react';
 import 'bulma/css/bulma.css';
 import '@fortawesome/fontawesome-free/css/all.css';
 
+import { Todo } from './types/Todo';
+import { getTodos } from './api';
 import { TodoList } from './components/TodoList';
 import { TodoFilter } from './components/TodoFilter';
 import { TodoModal } from './components/TodoModal';
 import { Loader } from './components/Loader';
 
+const debounce = (f: (query: string) => void, delay: number) => {
+  let timerId: number;
+
+  return (...args: string[]) => {
+    clearTimeout(timerId);
+
+    timerId = window.setTimeout(f, delay, ...args);
+  };
+};
+
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[] | null>(null);
+  const [selectedTodo, setSelectedTodo] = useState<Todo | null>(null);
+  const [filter, setFilter] = useState('all');
+  const [query, setQuery] = useState('');
+  const [appliedQuery, setAppliedQuery] = useState('');
+
+  useEffect(() => {
+    getTodos()
+      .then(todosList => setTodos(todosList));
+  }, [todos]);
+
+  const applyQuery = useCallback(
+    debounce(setAppliedQuery, 500), [],
+  );
+
+  const filteringTodos = (todosList: Todo[] | null, filterQuery: string) => {
+    if (!todosList) {
+      return null;
+    }
+
+    return todosList.filter(todo => {
+      switch (filter) {
+        case 'all':
+          return todo.title.toLowerCase().includes(filterQuery.toLowerCase());
+
+        case 'active':
+          return !todo.completed
+            && todo.title.toLowerCase().includes(filterQuery.toLowerCase());
+
+        case 'completed':
+          return todo.completed
+            && todo.title.toLowerCase().includes(filterQuery.toLowerCase());
+
+        default:
+          return todo;
+      }
+    });
+  };
+
+  const filteredTodos = filteringTodos(todos, appliedQuery);
```

Comment:
useMemo => now your arrray will have different link on every rerender and it will cause rerender of child component even if there will be no changes of filtered todos